This notebook aims in cleaning the dataset before starting using it to go over our questions. 

We use several techniques to clean the dataset along the notebook to facilitate the usage of the dataset.


We first import all the possibly needed modules.

In [1]:
import pandas as pd
import numpy as np
import datetime # to handle date/time attributes
from os import listdir # os is a module for interacting with the OS
from os.path import isfile, join # to verify file object, and concatenate paths
import glob # to find pathnames matching a specific pattern
import re # regular expressions :)

We first read the initial csv file into a dataframe that we call ***df***, we rename the columns to facilitate using it without spaces. Certain columns such as ***"DeathCountry", "DeathDate", "DeathCity", "Prize" and "Motivation"*** are dropped as we won't be using them.

The dataset consists of ***989 rows*** and ***14 columns***. 

In [2]:
df = pd.read_csv("nobel.csv")

# Display dataframe
df
df.rename(columns={'Organization City':'OrganizationCity',
                          'Organization Country':'OrganizationCountry',
                  'Organization Name':'OrganizationName','Death City':'DeathCity',
                  'Death Country':'DeathCountry','Death Date':'DeathDate',
                   'Full Name':'FullName'
                   
                  }, 
                 inplace=True)

df = df.drop({"DeathCountry", "DeathDate", "DeathCity", "Prize", "Motivation"}, axis=1)
#df = df.drop({"OrganizationCity", "OrganizationName"}, axis=1)



df



Year    Category Prize Share  Laureate ID Laureate Type  \
0    1901   Chemistry         1/1          160    Individual   
1    1901  Literature         1/1          569    Individual   
2    1901    Medicine         1/1          293    Individual   
3    1901       Peace         1/2          462    Individual   
4    1901       Peace         1/2          463    Individual   
5    1901     Physics         1/1            1    Individual   
6    1902   Chemistry         1/1          161    Individual   
7    1902  Literature         1/1          571    Individual   
8    1902    Medicine         1/1          294    Individual   
9    1902       Peace         1/2          464    Individual   
10   1902       Peace         1/2          465    Individual   
11   1902     Physics         1/2            2    Individual   
12   1902     Physics         1/2            3    Individual   
13   1903   Chemistry         1/1          162    Individual   
14   1903  Literature         1/1          572    Individual   
15   1903    Medicine         1/1          295    Individual   
16   1903       Peace         1/1          466    Individual   
17   1903     Physics         1/2            4    Individual   
18   1903     Physics         1/4            5    Individual   
19   1903     Physics         1/4            6    Individual   
20   1904   Chemistry         1/1          163    Individual   
21   1904  Literature         1/2          573    Individual   
22   1904  Literature         1/2          574    Individual   
23   1904    Medicine         1/1          296    Individual   
24   1904       Peace         1/1          467  Organization   
25   1904     Physics         1/1            8    Individual   
26   1905   Chemistry         1/1          164    Individual   
27   1905  Literature         1/1          575    Individual   
28   1905    Medicine         1/1          297    Individual   
29   1905       Peace         1/1          468    Individual   
..    ...         ...         ...          ...           ...   
939  2014       Peace         1/2          913    Individual   
940  2014       Peace         1/2          914    Individual   
941  2014     Physics         1/3          906    Individual   
942  2014     Physics         1/3          906    Individual   
943  2014     Physics         1/3          907    Individual   
944  2014     Physics         1/3          908    Individual   
945  2015   Chemistry         1/3          921    Individual   
946  2015   Chemistry         1/3          921    Individual   
947  2015   Chemistry         1/3          922    Individual   
948  2015   Chemistry         1/3          922    Individual   
949  2015   Chemistry         1/3          923    Individual   
950  2015   Economics         1/1          926    Individual   
951  2015  Literature         1/1          924    Individual   
952  2015    Medicine         1/4          916    Individual   
953  2015    Medicine         1/4          917    Individual   
954  2015    Medicine         1/2          918    Individual   
955  2015       Peace         1/1          925  Organization   
956  2015     Physics         1/2          919    Individual   
957  2015     Physics         1/2          920    Individual   
958  2016   Chemistry         1/3          931    Individual   
959  2016   Chemistry         1/3          932    Individual   
960  2016   Chemistry         1/3          933    Individual   
961  2016   Economics         1/2          935    Individual   
962  2016   Economics         1/2          936    Individual   
963  2016  Literature         1/1          937    Individual   
964  2016    Medicine         1/1          927    Individual   
965  2016       Peace         1/1          934    Individual   
966  2016     Physics         1/2          928    Individual   
967  2016     Physics         1/4          929    Individual   
968  2016     Physics         1/4          930    Individual   

                                    

The following cell is a trial to find a combination of a unique key pair to identify every record uniquely. 
We found that the combination of the columns ***'Laureate ID' and 'Year'*** is the one that gives ***937 different unique values***. We do not continue investigating the unique combination of the columns because we do not need a unique identifier for each record.

In [3]:
#WE DID NOT NEED TO MELT ANY COLUMNS SINCE THE DATA HAD ALREADY NO VALUES IN THE HEADERS OF THE COLUMNS

In [4]:
#df['Laureate ID'].nunique()
grp = df.groupby(['Laureate ID','Year']).size()
grp
#df.query('A > B')



Laureate ID  Year
1            1901    1
2            1902    1
3            1902    1
4            1903    1
5            1903    1
6            1903    1
             1911    1
8            1904    1
9            1905    1
10           1906    1
11           1907    1
12           1908    1
13           1909    1
14           1909    1
15           1910    1
16           1911    1
17           1912    1
18           1913    1
19           1914    1
20           1915    1
21           1915    1
22           1917    1
23           1918    1
24           1919    1
25           1920    1
26           1921    1
27           1922    1
28           1923    1
29           1924    1
30           1925    1
                    ..
908          2014    1
909          2014    1
910          2014    2
911          2014    1
912          2014    1
913          2014    1
914          2014    1
915          2014    1
916          2015    1
917          2015    1
918          2015    1
919          201

The count of missing values in each column is displayed.

In [5]:
#df['Laureate ID'].value_counts()
#tmp = df.groupby(['Laureate ID','Year','Organization Name', 'Organization City', 'Full Name']).size()
#tmp
#tmp.query(tmp.size()>1)

np.random.seed(0) 

missing_values_count_df = df.isnull().sum()
print(missing_values_count_df)

Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
Birth Date              29
Birth City              28
Birth Country           26
Sex                     26
OrganizationName       247
OrganizationCity       253
OrganizationCountry    253
dtype: int64


We search for the records that have a null value in the ***"OrganizationCountry" field *** but a not null ***"OrganizationCity" field***. We find only one record that we impute the ***"OrganizationCountry"*** by hand with **Tunisia** when **Tunis** was the ***OrganizationCity***.

In [6]:
a = df.query('OrganizationCountry.isnull() ', engine='python')
b = a.query(' OrganizationCity.notnull()', engine='python')
b

Year  Category Prize Share  Laureate ID Laureate Type  \
143  1928  Medicine         1/1          318    Individual   

                        FullName  Birth Date Birth City Birth Country   Sex  \
143  Charles Jules Henri Nicolle  1866-09-21      Rouen        France  Male   

     OrganizationName OrganizationCity OrganizationCountry  
143  Institut Pasteur            Tunis                 NaN

In [7]:
df.loc[df['OrganizationCity'] == 'Tunis', 'OrganizationCountry'] = 'Tunisia'

In [8]:
c = df.loc[df['OrganizationCity'] == 'Tunis']
c

Year  Category Prize Share  Laureate ID Laureate Type  \
143  1928  Medicine         1/1          318    Individual   

                        FullName  Birth Date Birth City Birth Country   Sex  \
143  Charles Jules Henri Nicolle  1866-09-21      Rouen        France  Male   

     OrganizationName OrganizationCity OrganizationCountry  
143  Institut Pasteur            Tunis             Tunisia

We search for the records that have a null value in the ***"OrganizationCountry" field *** with a not null ***"OrganizationName" field ***. This is done in order to find out which organization names need their ***"OrganizationCountry" field *** to be imputed. 

In [9]:

a = df.query('OrganizationCountry.isnull() ', engine='python')
b = a.query(' OrganizationName.notnull()', engine='python')
b

Year   Category Prize Share  Laureate ID Laureate Type  \
602  1988  Chemistry         1/3          270    Individual   
727  1999   Medicine         1/1          461    Individual   
775  2003  Chemistry         1/2          770    Individual   
830  2007   Medicine         1/3          811    Individual   
842  2008  Chemistry         1/3          831    Individual   
855  2009  Chemistry         1/3          842    Individual   
865  2009   Medicine         1/3          837    Individual   
899  2012  Chemistry         1/2          878    Individual   
924  2013   Medicine         1/3          885    Individual   
926  2013   Medicine         1/3          886    Individual   

                FullName  Birth Date                  Birth City  \
602   Johann Deisenhofer  1943-09-30                Zusamaltheim   
727        Günter Blobel  1936-05-21  Waltersdorf (Niegoslawice)   
775   Roderick MacKinnon  1956-02-19              Burlington, MA   
830    Mario R. Capecchi  1937-10-06                      Verona   
842       Roger Y. Tsien  1952-02-01                New York, NY   
855     Thomas A. Steitz  1940-08-23               Milwaukee, WI   
865      Jack W. Szostak  1952-11-09                      London   
899  Robert J. Lefkowitz  1943-04-15                New York, NY   
924    Randy W. Schekman  1948-12-30                St. Paul, MN   
926     Thomas C. Südhof  1955-12-22                   Göttingen   

                Birth Country   Sex                 OrganizationName  \
602                   Germany  Male  Howard Hughes Medical Institute   
727          Germany (Poland)  Male  Howard Hughes Medical Institute   
775  United States of America  Male  Howard Hughes Medical Institute   
830                     Italy  Male  Howard Hughes Medical Institute   
842  United States of America  Male  Howard Hughes Medical Institute   
855  United States of America  Male  Howard Hughes Medical Institute   
865            United Kingdom  Male  Howard Hughes Medical Institute   
899  United States of America  Male  Howard Hughes Medical Institute   
924  United States of America  Male  Howard Hughes Medical Institute   
926                   Germany  Male  Howard Hughes Medical Institute   

    OrganizationCity OrganizationCountry  
602              NaN                 NaN  
727              NaN                 NaN  
775              NaN                 NaN  
830              NaN                 NaN  
842              NaN                 NaN  
855              NaN                 NaN  
865              NaN                 NaN  
899              NaN                 NaN  
924              NaN                 NaN  
926              NaN                 NaN

According to the query result above, the organization with the name 'Howard Hughes Medical Institute' needs to be imputed. We try to find a record with the same organization name that has its ***"OrganizationCity" field *** not null.

In [10]:
Organization = df.loc[df['OrganizationName'] == 'Howard Hughes Medical Institute']
Organization = Organization.query(' OrganizationCity.notnull()', engine='python')
Organization

Year   Category Prize Share  Laureate ID Laureate Type      FullName  \
947  2015  Chemistry         1/3          922    Individual  Paul Modrich   

     Birth Date Birth City             Birth Country   Sex  \
947  1946-06-13  Raton, NM  United States of America  Male   

                    OrganizationName OrganizationCity  \
947  Howard Hughes Medical Institute       Durham, NC   

          OrganizationCountry  
947  United States of America

In [11]:
cityName = Organization['OrganizationCity'].values[0]
countryName = Organization['OrganizationCountry'].values[0]
print(cityName)

Durham, NC


We impute all records with the previously mentioned missing attribute with the results of the previous query.

In [12]:
df.loc[df['OrganizationName'] == 'Howard Hughes Medical Institute', 'OrganizationCountry'] = countryName
df.loc[df['OrganizationName'] == 'Howard Hughes Medical Institute', 'OrganizationCity'] = cityName

The below query shows the result of the records after imputing them. 

In [13]:
df.loc[df['OrganizationName'] == 'Howard Hughes Medical Institute']

Year   Category Prize Share  Laureate ID Laureate Type  \
602  1988  Chemistry         1/3          270    Individual   
727  1999   Medicine         1/1          461    Individual   
775  2003  Chemistry         1/2          770    Individual   
830  2007   Medicine         1/3          811    Individual   
842  2008  Chemistry         1/3          831    Individual   
855  2009  Chemistry         1/3          842    Individual   
865  2009   Medicine         1/3          837    Individual   
899  2012  Chemistry         1/2          878    Individual   
924  2013   Medicine         1/3          885    Individual   
926  2013   Medicine         1/3          886    Individual   
947  2015  Chemistry         1/3          922    Individual   

                FullName  Birth Date                  Birth City  \
602   Johann Deisenhofer  1943-09-30                Zusamaltheim   
727        Günter Blobel  1936-05-21  Waltersdorf (Niegoslawice)   
775   Roderick MacKinnon  1956-02-19              Burlington, MA   
830    Mario R. Capecchi  1937-10-06                      Verona   
842       Roger Y. Tsien  1952-02-01                New York, NY   
855     Thomas A. Steitz  1940-08-23               Milwaukee, WI   
865      Jack W. Szostak  1952-11-09                      London   
899  Robert J. Lefkowitz  1943-04-15                New York, NY   
924    Randy W. Schekman  1948-12-30                St. Paul, MN   
926     Thomas C. Südhof  1955-12-22                   Göttingen   
947         Paul Modrich  1946-06-13                   Raton, NM   

                Birth Country   Sex                 OrganizationName  \
602                   Germany  Male  Howard Hughes Medical Institute   
727          Germany (Poland)  Male  Howard Hughes Medical Institute   
775  United States of America  Male  Howard Hughes Medical Institute   
830                     Italy  Male  Howard Hughes Medical Institute   
842  United States of America  Male  Howard Hughes Medical Institute   
855  United States of America  Male  Howard Hughes Medical Institute   
865            United Kingdom  Male  Howard Hughes Medical Institute   
899  United States of America  Male  Howard Hughes Medical Institute   
924  United States of America  Male  Howard Hughes Medical Institute   
926                   Germany  Male  Howard Hughes Medical Institute   
947  United States of America  Male  Howard Hughes Medical Institute   

    OrganizationCity       OrganizationCountry  
602       Durham, NC  United States of America  
727       Durham, NC  United States of America  
775       Durham, NC  United States of America  
830       Durham, NC  United States of America  
842       Durham, NC  United States of America  
855       Durham, NC  United States of America  
865       Durham, NC  United States of America  
899       Durham, NC  United States of America  
924       Durham, NC  United States of America  
926       Durham, NC  United States of America  
947       Durham, NC  United States of America

After the above imputation, we check that the number of missing records in the ***"OrganizationCity" field *** and the ***"OrganizationCountry" field *** are reduced.

In [14]:
missing_values_count_df = df.isnull().sum()
print(missing_values_count_df)

Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
Birth Date              29
Birth City              28
Birth Country           26
Sex                     26
OrganizationName       247
OrganizationCity       243
OrganizationCountry    242
dtype: int64


This step is to further make sure that there are no records with ***"OrganizationName" field *** not null and the ***"OrganizationCountry" field *** null.


In [15]:
a = df.query('OrganizationCountry.isnull() ', engine='python')
b = a.query(' OrganizationName.notnull()', engine='python')
b

Empty DataFrame
Columns: [Year, Category, Prize Share, Laureate ID, Laureate Type, FullName, Birth Date, Birth City, Birth Country, Sex, OrganizationName, OrganizationCity, OrganizationCountry]
Index: []

In this step, we tried to find whether there is a record with the ***"OrganizationName" field *** as 'University of Delaware' and ***"OrganizationCity" field *** not null so we could impute the missing records.

In [16]:
df.loc[df['OrganizationName'] == 'University of Delaware']

Year   Category Prize Share  Laureate ID Laureate Type         FullName  \
871  2010  Chemistry         1/3          851    Individual  Richard F. Heck   

     Birth Date       Birth City             Birth Country   Sex  \
871  1931-08-15  Springfield, MA  United States of America  Male   

           OrganizationName OrganizationCity       OrganizationCountry  
871  University of Delaware              NaN  United States of America

The above query result shows that there are no imputations that can be made.

In [17]:
missing_values_count_df = df.isnull().sum()
print(missing_values_count_df)

Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
Birth Date              29
Birth City              28
Birth Country           26
Sex                     26
OrganizationName       247
OrganizationCity       243
OrganizationCountry    242
dtype: int64


We use an external dataset "nobel_prize_by_winner" in order to help us with some imputations in our original dataset.

In [18]:
external_df1 =  pd.read_csv("./external data/nobel_prize_by_winner.csv")
external_df1
                            

id                                          firstname  \
0    846                                             Elinor   
1    846                                             Elinor   
2    783                                       Wangari Muta   
3    230                                   Dorothy Crowfoot   
4    918                                             Youyou   
5    428                                            Barbara   
6    773                                             Shirin   
7    597                                             Grazia   
8    615                                           Gabriela   
9    782                                           Elfriede   
10   553                                   Aung San Suu Kyi   
11   640                                              Nelly   
12   924                                           Svetlana   
13   610                                              Pearl   
14   673                                            Wislawa   
15   843                                             Ada E.   
16   413                                            Rosalyn   
17   836                                           Carol W.   
18   835                                       Elizabeth H.   
19    79                                              Maria   
20   453                                         Christiane   
21   904                                          May-Britt   
22   435                                               Rita   
23   438                                        Gertrude B.   
24   775                                           Linda B.   
25   824                                          Fran�oise   
26   344                                      Gerty Theresa   
27   869                                              Ellen   
28   870                                             Leymah   
29   871                                           Tawakkol   
..   ...                                                ...   
942  500                                               Carl   
943  518                                          Philip J.   
944  516                                      Lester Bowles   
945  519                                        Albert John   
946  925                          National Dialogue Quartet   
947  893  Organisation for the Prohibition of Chemical W...   
948  881                                European Union (EU)   
949  818   Intergovernmental Panel on Climate Change (IPCC)   
950  810                                       Grameen Bank   
951  561   Pugwash Conferences on Science and World Affairs   
952  797          International Atomic Energy Agency (IAEA)   
953  748                              United Nations (U.N.)   
954  564     International Campaign to Ban Landmines (ICBL)   
955  568                           M_decins Sans Fronti�res   
956  467  Institut de droit international (Institute of ...   
957  477  Bureau international permanent de la Paix (Per...   
958  482  Comit_ international de la Croix Rouge (Intern...   
959  482  Comit_ international de la Croix Rouge (Intern...   
960  482  Comit_ international de la Croix Rouge (Intern...   
961  503  Office international Nansen pour les R_fugi_s ...   
962  508              Friends Service Council (The Quakers)   
963  509   American Friends Service Committee (The Quakers)   
964  515  Office of the United Nations High Commissioner...   
965  515  Office of the United Nations High Commissioner...   
966  523  Ligue des Soci_t_s de la Croix-Rouge (League o...   
967  525            United Nations Children's Fund (UNICEF)   
968  527         International Labour Organization (I.L.O.)   
969  537                              Amnesty International   
970  547  International Physicians for the Prevention of...   
971  550                 United Nations Peacekeeping Forces   

               surname        born        died  \
0               Ostrom    8/7/1933   6/12/2012   

Then, we check the number of records with missing values to compare both datasets.

In [19]:
missing_values_count_df1 = external_df1.isnull().sum()
print(df.shape[0])

print(missing_values_count_df)

print(external_df1.shape[0])

print(missing_values_count_df1)

969
Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
Birth Date              29
Birth City              28
Birth Country           26
Sex                     26
OrganizationName       247
OrganizationCity       243
OrganizationCountry    242
dtype: int64
972
id                     0
firstname              6
surname               36
born                  26
died                  26
borncountry           32
borncountrycode       32
borncity              34
diedcountry          371
diedcountrycode      372
diedcity             377
gender                 0
year                   6
category               6
overallmotivation    956
share                  6
motivation            94
name                 252
city                 258
country              258
dtype: int64


Due to the results above, it can be seen that the new dataset has no missing values in the ***"Gender" field ***. Therefore, we merge both datasets in order to make use of this field to impute the missing sex values.

In [20]:
#new_df = pd.merge(df, external_df1, left_on="Laureate ID", right_on="id")
#new_df['Laureate ID',]
#new_df[['Laureate ID','id']]
new_df = pd.merge(df, external_df1, left_on=["Laureate ID","OrganizationName","OrganizationCity","OrganizationCountry"], right_on=["id","name","city","country"], how = 'left')
new_df
#new_df.groupby(['Laureate ID','Year']).size()
#new_df.loc[new_df['Laureate ID'] == 921]


Year    Category Prize Share  Laureate ID Laureate Type  \
0    1901   Chemistry         1/1          160    Individual   
1    1901  Literature         1/1          569    Individual   
2    1901    Medicine         1/1          293    Individual   
3    1901       Peace         1/2          462    Individual   
4    1901       Peace         1/2          463    Individual   
5    1901     Physics         1/1            1    Individual   
6    1902   Chemistry         1/1          161    Individual   
7    1902  Literature         1/1          571    Individual   
8    1902    Medicine         1/1          294    Individual   
9    1902       Peace         1/2          464    Individual   
10   1902       Peace         1/2          465    Individual   
11   1902     Physics         1/2            2    Individual   
12   1902     Physics         1/2            3    Individual   
13   1903   Chemistry         1/1          162    Individual   
14   1903  Literature         1/1          572    Individual   
15   1903    Medicine         1/1          295    Individual   
16   1903       Peace         1/1          466    Individual   
17   1903     Physics         1/2            4    Individual   
18   1903     Physics         1/4            5    Individual   
19   1903     Physics         1/4            6    Individual   
20   1904   Chemistry         1/1          163    Individual   
21   1904  Literature         1/2          573    Individual   
22   1904  Literature         1/2          574    Individual   
23   1904    Medicine         1/1          296    Individual   
24   1904       Peace         1/1          467  Organization   
25   1904     Physics         1/1            8    Individual   
26   1905   Chemistry         1/1          164    Individual   
27   1905  Literature         1/1          575    Individual   
28   1905    Medicine         1/1          297    Individual   
29   1905       Peace         1/1          468    Individual   
..    ...         ...         ...          ...           ...   
947  2014       Peace         1/2          913    Individual   
948  2014       Peace         1/2          914    Individual   
949  2014     Physics         1/3          906    Individual   
950  2014     Physics         1/3          906    Individual   
951  2014     Physics         1/3          907    Individual   
952  2014     Physics         1/3          908    Individual   
953  2015   Chemistry         1/3          921    Individual   
954  2015   Chemistry         1/3          921    Individual   
955  2015   Chemistry         1/3          922    Individual   
956  2015   Chemistry         1/3          922    Individual   
957  2015   Chemistry         1/3          923    Individual   
958  2015   Economics         1/1          926    Individual   
959  2015  Literature         1/1          924    Individual   
960  2015    Medicine         1/4          916    Individual   
961  2015    Medicine         1/4          917    Individual   
962  2015    Medicine         1/2          918    Individual   
963  2015       Peace         1/1          925  Organization   
964  2015     Physics         1/2          919    Individual   
965  2015     Physics         1/2          920    Individual   
966  2016   Chemistry         1/3          931    Individual   
967  2016   Chemistry         1/3          932    Individual   
968  2016   Chemistry         1/3          933    Individual   
969  2016   Economics         1/2          935    Individual   
970  2016   Economics         1/2          936    Individual   
971  2016  Literature         1/1          937    Individual   
972  2016    Medicine         1/1          927    Individual   
973  2016       Peace         1/1          934    Individual   
974  2016     Physics         1/2          928    Individual   
975  2016     Physics         1/4          929    Individual   
976  2016     Physics         1/4          930    Individual   

                                    

To impute the records with the missing Sex values, we find the records with ***"Sex" field *** as null and we display their Laureate ID.

In [21]:
missing_sex_df = df[['Laureate ID','Sex']]
missing_sex_df = df.query('Sex.isnull() ', engine='python')
missing_sex_df = missing_sex_df[['Laureate ID','Sex']]
missing_sex_df = missing_sex_df[['Laureate ID']]
missing_sex_df



Laureate ID
24           467
61           477
90           482
206          503
222          482
244          508
245          509
295          515
365          482
366          523
383          525
416          527
498          537
543          515
578          547
610          550
682          561
705          564
728          568
754          748
808          797
821          810
833          818
909          881
927          893
955          925

We merge the previous dataframe with the external dataset on ***"ID" field *** and ***"Laureate ID" field *** and on the ***"Gender" field *** and ***"Sex" field ***. We rename the Gender column as Sex and display only the Laureate ID and Sex columns.

In [22]:
sex_df1 = external_df1[['id','gender']]

sex_new_df = pd.merge(missing_sex_df, sex_df1, how='inner', left_on='Laureate ID', right_on='id')
sex_new_df.shape[0]

sex_new_df = sex_new_df.drop_duplicates()
sex_new_df.shape[0]

sex_new_df.rename(columns={'gender':'Sex'}, inplace=True)

sex_new_df = sex_new_df[['Laureate ID', 'Sex']]
sex_new_df


Laureate ID  Sex
0           467  org
1           477  org
2           482  org
11          503  org
12          508  org
13          509  org
14          515  org
18          523  org
19          525  org
20          527  org
21          537  org
22          547  org
23          550  org
24          561  org
25          564  org
26          568  org
27          748  org
28          797  org
29          810  org
30          818  org
31          881  org
32          893  org
33          925  org

After obtaining the result above, we can see that the only records with missing Sex values are the organizations. We then merge the original dataframe with the newly imputed one to have a final ***"Sex" field *** with no missing values.


In [23]:
new_df = pd.merge(df, sex_new_df, how='left', left_on='Laureate ID', right_on='Laureate ID')
new_df['Sex_x'] = np.where(new_df['Sex_x'].isnull(), new_df['Sex_y'], new_df['Sex_x'])
new_df.loc[new_df['Sex_x'] == 'org']

new_df.rename(columns={'Sex_x':'Sex'}, inplace=True)
new_df = new_df.drop("Sex_y", axis=1)
new_df



Year    Category Prize Share  Laureate ID Laureate Type  \
0    1901   Chemistry         1/1          160    Individual   
1    1901  Literature         1/1          569    Individual   
2    1901    Medicine         1/1          293    Individual   
3    1901       Peace         1/2          462    Individual   
4    1901       Peace         1/2          463    Individual   
5    1901     Physics         1/1            1    Individual   
6    1902   Chemistry         1/1          161    Individual   
7    1902  Literature         1/1          571    Individual   
8    1902    Medicine         1/1          294    Individual   
9    1902       Peace         1/2          464    Individual   
10   1902       Peace         1/2          465    Individual   
11   1902     Physics         1/2            2    Individual   
12   1902     Physics         1/2            3    Individual   
13   1903   Chemistry         1/1          162    Individual   
14   1903  Literature         1/1          572    Individual   
15   1903    Medicine         1/1          295    Individual   
16   1903       Peace         1/1          466    Individual   
17   1903     Physics         1/2            4    Individual   
18   1903     Physics         1/4            5    Individual   
19   1903     Physics         1/4            6    Individual   
20   1904   Chemistry         1/1          163    Individual   
21   1904  Literature         1/2          573    Individual   
22   1904  Literature         1/2          574    Individual   
23   1904    Medicine         1/1          296    Individual   
24   1904       Peace         1/1          467  Organization   
25   1904     Physics         1/1            8    Individual   
26   1905   Chemistry         1/1          164    Individual   
27   1905  Literature         1/1          575    Individual   
28   1905    Medicine         1/1          297    Individual   
29   1905       Peace         1/1          468    Individual   
..    ...         ...         ...          ...           ...   
939  2014       Peace         1/2          913    Individual   
940  2014       Peace         1/2          914    Individual   
941  2014     Physics         1/3          906    Individual   
942  2014     Physics         1/3          906    Individual   
943  2014     Physics         1/3          907    Individual   
944  2014     Physics         1/3          908    Individual   
945  2015   Chemistry         1/3          921    Individual   
946  2015   Chemistry         1/3          921    Individual   
947  2015   Chemistry         1/3          922    Individual   
948  2015   Chemistry         1/3          922    Individual   
949  2015   Chemistry         1/3          923    Individual   
950  2015   Economics         1/1          926    Individual   
951  2015  Literature         1/1          924    Individual   
952  2015    Medicine         1/4          916    Individual   
953  2015    Medicine         1/4          917    Individual   
954  2015    Medicine         1/2          918    Individual   
955  2015       Peace         1/1          925  Organization   
956  2015     Physics         1/2          919    Individual   
957  2015     Physics         1/2          920    Individual   
958  2016   Chemistry         1/3          931    Individual   
959  2016   Chemistry         1/3          932    Individual   
960  2016   Chemistry         1/3          933    Individual   
961  2016   Economics         1/2          935    Individual   
962  2016   Economics         1/2          936    Individual   
963  2016  Literature         1/1          937    Individual   
964  2016    Medicine         1/1          927    Individual   
965  2016       Peace         1/1          934    Individual   
966  2016     Physics         1/2          928    Individual   
967  2016     Physics         1/4          929    Individual   
968  2016     Physics         1/4          930    Individual   

                                    

We check that the ***"Sex" field *** now has no missing values.

In [24]:
missing_values_count_df = new_df.isnull().sum()
print(missing_values_count_df)

print('\n')

missing_values_count_df1 = external_df1.isnull().sum()
print(missing_values_count_df1)

Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
Birth Date              29
Birth City              28
Birth Country           26
Sex                      0
OrganizationName       247
OrganizationCity       243
OrganizationCountry    242
dtype: int64


id                     0
firstname              6
surname               36
born                  26
died                  26
borncountry           32
borncountrycode       32
borncity              34
diedcountry          371
diedcountrycode      372
diedcity             377
gender                 0
year                   6
category               6
overallmotivation    956
share                  6
motivation            94
name                 252
city                 258
country              258
dtype: int64


We proceed to check if the ***"Birth Date" field *** can be imputed using the external dataset by displaying a dataframe with the ***"Birth Date" field *** and ***"Born" field ***. We discover that the same records of both datasets have missing values in both fields and hence, we cannot impute the ***"Birth Date" field ***.

In [25]:
new_df.rename(columns={'Birth Date':'BirthDate'}, inplace=True)
a = new_df.query('BirthDate.isnull() ', engine='python')
a

Year   Category Prize Share  Laureate ID Laureate Type  \
24   1904      Peace         1/1          467  Organization   
61   1910      Peace         1/1          477  Organization   
90   1917      Peace         1/1          482  Organization   
206  1938      Peace         1/1          503  Organization   
222  1944      Peace         1/1          482  Organization   
244  1947      Peace         1/2          508  Organization   
245  1947      Peace         1/2          509  Organization   
295  1954      Peace         1/1          515  Organization   
365  1963      Peace         1/2          482  Organization   
366  1963      Peace         1/2          523  Organization   
383  1965      Peace         1/1          525  Organization   
416  1969      Peace         1/1          527  Organization   
498  1977      Peace         1/1          537  Organization   
543  1981      Peace         1/1          515  Organization   
578  1985      Peace         1/1          547  Organization   
610  1988      Peace         1/1          550  Organization   
682  1995      Peace         1/2          561  Organization   
705  1997      Peace         1/2          564  Organization   
728  1999      Peace         1/1          568  Organization   
754  2001      Peace         1/2          748  Organization   
808  2005      Peace         1/2          797  Organization   
821  2006      Peace         1/2          810  Organization   
833  2007      Peace         1/2          818  Organization   
853  2009  Chemistry         1/3          841    Individual   
894  2011    Physics         1/2          864    Individual   
895  2011    Physics         1/2          864    Individual   
909  2012      Peace         1/1          881  Organization   
927  2013      Peace         1/1          893  Organization   
955  2015      Peace         1/1          925  Organization   

                                              FullName BirthDate  \
24   Institut de droit international (Institute of ...       NaN   
61   Bureau international permanent de la Paix (Per...       NaN   
90   Comité international de la Croix Rouge (Intern...       NaN   
206  Office international Nansen pour les Réfugiés ...       NaN   
222  Comité international de la Croix Rouge (Intern...       NaN   
244              Friends Service Council (The Quakers)       NaN   
245   American Friends Service Committee (The Quakers)       NaN   
295  Office of the United Nations High Commissioner...       NaN   
365  Comité international de la Croix Rouge (Intern...       NaN   
366  Ligue des Sociétés de la Croix-Rouge (League o...       NaN   
383            United Nations Children's Fund (UNICEF)       NaN   
416         International Labour Organization (I.L.O.)       NaN   
498                              Amnesty International       NaN   
543  Office of the United Nations High Commissioner...       NaN   
578  International Physicians for the Prevention of...       NaN   
610                 United Nations Peacekeeping Forces       NaN   
682   Pugwash Conferences on Science and World Affairs       NaN   
705     International Campaign to Ban Landmines (ICBL)       NaN   
728                           Médecins Sans Frontières       NaN   
754                              United Nations (U.N.)       NaN   
808          International Atomic Energy Agency (IAEA)       NaN   
821                                       Grameen Bank       NaN   
833   Intergovernmental Panel on Climate Change (IPCC)       NaN   
853                           Venkatraman Ramakrishnan       NaN   
894                                    Saul Perlmutter       NaN   
895                                    Saul Perlmutter       NaN   
909                                European Union (EU)       NaN   
927  Organisation for the Prohibition of Chemical W...       NaN   
955                          National Dialogue Quartet       NaN   

                  Birth City             Birth Country   Sex  \
24       

In [26]:
df2 = pd.merge(a, external_df1, how='inner', left_on='Laureate ID', right_on='id')
df2[['BirthDate', 'born']]

BirthDate        born
0        NaN         NaN
1        NaN         NaN
2        NaN         NaN
3        NaN         NaN
4        NaN         NaN
5        NaN         NaN
6        NaN         NaN
7        NaN         NaN
8        NaN         NaN
9        NaN         NaN
10       NaN         NaN
11       NaN         NaN
12       NaN         NaN
13       NaN         NaN
14       NaN         NaN
15       NaN         NaN
16       NaN         NaN
17       NaN         NaN
18       NaN         NaN
19       NaN         NaN
20       NaN         NaN
21       NaN         NaN
22       NaN         NaN
23       NaN         NaN
24       NaN         NaN
25       NaN         NaN
26       NaN         NaN
27       NaN         NaN
28       NaN         NaN
29       NaN         NaN
30       NaN         NaN
31       NaN  0000/00/00
32       NaN  0000/00/00
33       NaN  0000/00/00
34       NaN  0000/00/00
35       NaN  0000/00/00
36       NaN         NaN
37       NaN         NaN
38       NaN         NaN

In [27]:
a = new_df.query('OrganizationName.isnull() ', engine='python')
df2 = pd.merge(a, external_df1, how='inner', left_on='Laureate ID', right_on='id')
df2[['OrganizationName', 'name']]
df3 = df2.query('name.notnull() ', engine='python')
df3[['OrganizationName', 'name']]

OrganizationName                 name
8                NaN  Sorbonne University
114              NaN     Mainz University

In [28]:
a = new_df.query('OrganizationCity.isnull() ', engine='python')
df2 = pd.merge(a, external_df1, how='inner', left_on='Laureate ID', right_on='id')
df2[['OrganizationCity', 'city']]
df3 = df2.query('city.notnull() ', engine='python')
df3[['OrganizationCity', 'city']]

OrganizationCity   city
8              NaN  Paris

In [29]:
a = new_df.query('OrganizationCountry.isnull() ', engine='python')
df2 = pd.merge(a, external_df1, how='inner', left_on='Laureate ID', right_on='id')
df2[['OrganizationCountry', 'country']]
df3 = df2.query('city.notnull() ', engine='python')
df3[['OrganizationCountry', 'country']]

OrganizationCountry country
8                 NaN  France

In [30]:
tmp = new_df.query('OrganizationName.isnull() ', engine='python')
tmp2 = tmp[['Category', 'OrganizationName']]
tmp2.groupby(['Category']).size()

Category
Chemistry       2
Economics       2
Literature    113
Medicine        2
Peace         126
Physics         2
dtype: int64

In [31]:
tmp = new_df.query('OrganizationName.notnull() ', engine='python')
tmp2 = tmp[['Category', 'OrganizationName']]
tmp2.groupby(['Category']).size()

Category
Chemistry    192
Economics     81
Medicine     225
Peace          4
Physics      220
dtype: int64

After researching on some of the Nobel Prize Winners in the Literature category, and since there were no records with the Literature category having Organization Name as not null, we came to a conclusion that they are always individuals, not organizations. Therefore, we impute these records with Not Applicable in the ***"Organization Name" field *** and ***"Organization Country" field ***.

In [32]:
mask = ((new_df['OrganizationName'].isnull()) & (new_df['Category'] == 'Literature'))
new_df['OrganizationName'] = new_df['OrganizationName'].mask(mask, 'Not Applicable')


#mask = ((new_df['OrganizationName'].isnull()) & (new_df['Category'] == 'Peace'))
#new_df['OrganizationName'] = new_df['OrganizationName'].mask(mask, 'Not Applicable')

#mask = ((new_df['OrganizationCountry'].isnull()) & (new_df['Category'] == 'Peace'))
#new_df['OrganizationCountry'] = new_df['OrganizationCountry'].mask(mask, 'Not Applicable')

mask = ((new_df['OrganizationCountry'].isnull()) & (new_df['Category'] == 'Literature'))
new_df['OrganizationCountry'] = new_df['OrganizationCountry'].mask(mask, 'Not Applicable')
                                                         

    
new_df = new_df.drop({"OrganizationCity"}, axis=1)
    
    
new_df


Year    Category Prize Share  Laureate ID Laureate Type  \
0    1901   Chemistry         1/1          160    Individual   
1    1901  Literature         1/1          569    Individual   
2    1901    Medicine         1/1          293    Individual   
3    1901       Peace         1/2          462    Individual   
4    1901       Peace         1/2          463    Individual   
5    1901     Physics         1/1            1    Individual   
6    1902   Chemistry         1/1          161    Individual   
7    1902  Literature         1/1          571    Individual   
8    1902    Medicine         1/1          294    Individual   
9    1902       Peace         1/2          464    Individual   
10   1902       Peace         1/2          465    Individual   
11   1902     Physics         1/2            2    Individual   
12   1902     Physics         1/2            3    Individual   
13   1903   Chemistry         1/1          162    Individual   
14   1903  Literature         1/1          572    Individual   
15   1903    Medicine         1/1          295    Individual   
16   1903       Peace         1/1          466    Individual   
17   1903     Physics         1/2            4    Individual   
18   1903     Physics         1/4            5    Individual   
19   1903     Physics         1/4            6    Individual   
20   1904   Chemistry         1/1          163    Individual   
21   1904  Literature         1/2          573    Individual   
22   1904  Literature         1/2          574    Individual   
23   1904    Medicine         1/1          296    Individual   
24   1904       Peace         1/1          467  Organization   
25   1904     Physics         1/1            8    Individual   
26   1905   Chemistry         1/1          164    Individual   
27   1905  Literature         1/1          575    Individual   
28   1905    Medicine         1/1          297    Individual   
29   1905       Peace         1/1          468    Individual   
..    ...         ...         ...          ...           ...   
939  2014       Peace         1/2          913    Individual   
940  2014       Peace         1/2          914    Individual   
941  2014     Physics         1/3          906    Individual   
942  2014     Physics         1/3          906    Individual   
943  2014     Physics         1/3          907    Individual   
944  2014     Physics         1/3          908    Individual   
945  2015   Chemistry         1/3          921    Individual   
946  2015   Chemistry         1/3          921    Individual   
947  2015   Chemistry         1/3          922    Individual   
948  2015   Chemistry         1/3          922    Individual   
949  2015   Chemistry         1/3          923    Individual   
950  2015   Economics         1/1          926    Individual   
951  2015  Literature         1/1          924    Individual   
952  2015    Medicine         1/4          916    Individual   
953  2015    Medicine         1/4          917    Individual   
954  2015    Medicine         1/2          918    Individual   
955  2015       Peace         1/1          925  Organization   
956  2015     Physics         1/2          919    Individual   
957  2015     Physics         1/2          920    Individual   
958  2016   Chemistry         1/3          931    Individual   
959  2016   Chemistry         1/3          932    Individual   
960  2016   Chemistry         1/3          933    Individual   
961  2016   Economics         1/2          935    Individual   
962  2016   Economics         1/2          936    Individual   
963  2016  Literature         1/1          937    Individual   
964  2016    Medicine         1/1          927    Individual   
965  2016       Peace         1/1          934    Individual   
966  2016     Physics         1/2          928    Individual   
967  2016     Physics         1/4          929    Individual   
968  2016     Physics         1/4          930    Individual   

                                    

In [33]:
new_df

Year    Category Prize Share  Laureate ID Laureate Type  \
0    1901   Chemistry         1/1          160    Individual   
1    1901  Literature         1/1          569    Individual   
2    1901    Medicine         1/1          293    Individual   
3    1901       Peace         1/2          462    Individual   
4    1901       Peace         1/2          463    Individual   
5    1901     Physics         1/1            1    Individual   
6    1902   Chemistry         1/1          161    Individual   
7    1902  Literature         1/1          571    Individual   
8    1902    Medicine         1/1          294    Individual   
9    1902       Peace         1/2          464    Individual   
10   1902       Peace         1/2          465    Individual   
11   1902     Physics         1/2            2    Individual   
12   1902     Physics         1/2            3    Individual   
13   1903   Chemistry         1/1          162    Individual   
14   1903  Literature         1/1          572    Individual   
15   1903    Medicine         1/1          295    Individual   
16   1903       Peace         1/1          466    Individual   
17   1903     Physics         1/2            4    Individual   
18   1903     Physics         1/4            5    Individual   
19   1903     Physics         1/4            6    Individual   
20   1904   Chemistry         1/1          163    Individual   
21   1904  Literature         1/2          573    Individual   
22   1904  Literature         1/2          574    Individual   
23   1904    Medicine         1/1          296    Individual   
24   1904       Peace         1/1          467  Organization   
25   1904     Physics         1/1            8    Individual   
26   1905   Chemistry         1/1          164    Individual   
27   1905  Literature         1/1          575    Individual   
28   1905    Medicine         1/1          297    Individual   
29   1905       Peace         1/1          468    Individual   
..    ...         ...         ...          ...           ...   
939  2014       Peace         1/2          913    Individual   
940  2014       Peace         1/2          914    Individual   
941  2014     Physics         1/3          906    Individual   
942  2014     Physics         1/3          906    Individual   
943  2014     Physics         1/3          907    Individual   
944  2014     Physics         1/3          908    Individual   
945  2015   Chemistry         1/3          921    Individual   
946  2015   Chemistry         1/3          921    Individual   
947  2015   Chemistry         1/3          922    Individual   
948  2015   Chemistry         1/3          922    Individual   
949  2015   Chemistry         1/3          923    Individual   
950  2015   Economics         1/1          926    Individual   
951  2015  Literature         1/1          924    Individual   
952  2015    Medicine         1/4          916    Individual   
953  2015    Medicine         1/4          917    Individual   
954  2015    Medicine         1/2          918    Individual   
955  2015       Peace         1/1          925  Organization   
956  2015     Physics         1/2          919    Individual   
957  2015     Physics         1/2          920    Individual   
958  2016   Chemistry         1/3          931    Individual   
959  2016   Chemistry         1/3          932    Individual   
960  2016   Chemistry         1/3          933    Individual   
961  2016   Economics         1/2          935    Individual   
962  2016   Economics         1/2          936    Individual   
963  2016  Literature         1/1          937    Individual   
964  2016    Medicine         1/1          927    Individual   
965  2016       Peace         1/1          934    Individual   
966  2016     Physics         1/2          928    Individual   
967  2016     Physics         1/4          929    Individual   
968  2016     Physics         1/4          930    Individual   

                                    

In [34]:
new_df.to_csv('new_nobel.csv')

missing_values_count_df = new_df.isnull().sum()
print(missing_values_count_df)
print('\n')
print(new_df.nunique())

Year                     0
Category                 0
Prize Share              0
Laureate ID              0
Laureate Type            0
FullName                 0
BirthDate               29
Birth City              28
Birth Country           26
Sex                      0
OrganizationName       134
OrganizationCountry    129
dtype: int64


Year                   113
Category                 6
Prize Share              4
Laureate ID            904
Laureate Type            2
FullName               904
BirthDate              868
Birth City             601
Birth Country          121
Sex                      3
OrganizationName       316
OrganizationCountry     31
dtype: int64
